rfcs/20190305-modular-tensorflow.md

```diff
@@ -0,0 +1,481 @@
+
+Currently, this means that all such hardware developers need to check in their code into the [main tensorflow repository](http://github.com/tensorflow/tensorflow). This means that all the changes are required to go through TF team's review. This can make merging support for new hardware very very difficult.
+
+#### Long PR review queue
+
+TensorFlow is a very successful opensource project. It is the [4th most forked](https://github.com/search?o=desc&q=stars:%3E1&s=forks&type=Repositories) and [5th most starred](https://github.com/search?q=stars%3A%3E0&s=stars&type=Repositories) project on github. This also means that TF receives quite a lot of opensource contributions. The TensorFlow team has to review all contributions to TensorFlow itself. This creates a bottleneck for merging the changes to the main repository.
+
+#### Flexibility for collaborators
+
+Currently, any partner or contributor that would like to work with us are subject to all the rules within of the main repository. Some of these can be relaxed through modularization, where work can happen in a separate repository.
+
+#### Large TF support matrix
+
+The TF support story is a unique beast. Our support matrix has a lot of orthogonal dimensions. Below are some of the more prominent dimensions:
+
+*   Environment (google3, opensource)
+*   Operating system (Linux, windows, macos, mobile)
+*   Architecture (x86, arm, ppc64, …)
+*   Accelerator (CPU, GPU,  TPU)
+*   Compiler (GCC, Clang, MSVC)
+*   Python version (2, 3.4, 3.5, 3.6,  3.7)
+
+More can be added to this list where we have to rebuild TensorFlow to support different network architectures, CUDA versions, SIMD instruction sets, etc.
+
+Having a monolithic repository means we need to rebuild all of our code for all of these different combinations. However, it makes no sense to rebuild all of our C++ code if the only difference is the Python version. Or rebuild all of our CPU kernels for different CUDA versions. Modularizing our code means we only need to rebuild and test the modules that are directly impacted by the dimensions we are changing in the support matrix.
+
+
+## Overview
+
+This project aims to split the TensorFlow codebase into **smaller, more focused**, repositories that can be released and managed separately. These modules will talk to each other using **well defined APIs** that guarantee backwards compatibility. Thanks to these APIs, these modules will be **managed/owned/released independently**. There will be different strategies to break apart pieces based on the languages, but below summarizes the approach for C++ and Python:
+
+
+![alt_text](20190305-modular-tensorflow/big_picture.png "Overview of modular TensorFlow")
+
+A summary of the above is:
+
+
+
+*   Core TF functionality will be implemented in C++
+*   Core TF functionality can be extended using shared objects.
+*   On top of the core C++ libraries, we will have the language bindings (Using the C API)
```

Comment:
As discussed in [this developer thread](https://groups.google.com/a/tensorflow.org/forum/?utm_medium=email&utm_source=footer#!msg/developers/TuN9YDVVOxE/bUP_KwZQCgAJ), it is still not very clear to me how protobufs will be exposed by the C API.

For example, the language bindings will need to retrieve the list of available operations, which is currently represented by a list of `OpDef` described in [opdef.proto](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/framework/op_def.proto). How will that information cross the C API? 
1. by returning the protobufs as a series of bytes that should be deserialized by the bindings, like in [TF_GetAllRegisteredKernels](https://github.com/tensorflow/tensorflow/blob/1f18aeee0fe45207eb27b4a3d0a6ebbaae165c78/tensorflow/c/c_api.h#L1724)?
2. by exposing a snapshot of the protobufs as a text file that should be parsed by bindings, like in [ops.pbtxt](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/ops/ops.pbtxt)?
3. by wrapping protobuf messages in a C structure exposed in the API?

In other words, can a `.proto` file of TensorFlow be part of the contract with its modules or should they be used only internally by the core?

Reply:
Our solution to this will be 3.
We would like to remove all protobufs from our APIs.
We will push for using protobufs for only serialization formats, for RPCs or data we would store on disk.

Reply:
@gunan It is not clear how to add new operators to a module plugin. Currently, we could create new operators through `OpDef` and optionally set visibility hidden. We have been doing this for MKL-DNN quantization work.

Reply:
The idea is to use a mechanism similar to custom ops, but simplified. With a C ABI, we will be able to make custom ops much easier to build and distribute.

Reply:
Quick question, how will the TensorFlow libraries be published to be used by language bindings?

Right now, pretty much all bindings have their own distribution (PiP package, Maven artifacts...) that bundles both the client and the TF core libraries. But with module TF, will the core libraries built, tested and published by Google so the clients just need to pick them for distribution?